README.md
# SDP2019Team23 from UMass Amherst ECE 
Team Website: http://www.ecs.umass.edu/ece/sdp/sdp19/team23/node_modules/startbootstrap-stylish-portfolio/
## For Vistor
All detailed tasks will be recorded as issues in repo. Each project board represent a weekly sprint. All additional can be found on Wiki page. Some documentation won't be availabe until it is reviewed. **Feel free to reach out any of us to discuss this project**, we are intending to opensource this May,2019 when this project is finished.

## Team members:
Gongtao Yang <gyang@umass.edu>,    ToddHuaji, Avoidance Lead, Manager(Oct-present);  
Conley Gao <Haogao@umass.edu>,   ConleyGao, Proximity Lead, Manager(Sept-Oct);  
Kai Qian <kaiqian@umass.edu>,  kaiqiann, Sensor Driver Lead;  
Anthony Abou-Francis <aaboufrancis@umass.edu>, tony5598, Actions Handlers Lead

### Faculty Advisor: Prof. Michal Zink <zink@ecs.umass.edu>


ArduCopter/ReleaseNotes.txt
ArduPilot Copter Release Notes:
------------------------------------------------------------------
<<<<<<< HEAD
Copter 3.6.5 24-Jan-2019 / 3.6.5-rc3 21-Jan-2019
=======
Copter 3.6.5-rc3 21-Jan-2019
>>>>>>> 8369de2c8092ecc66d35e97036bb6fb3902e3f24
Changes from 3.6.5-rc2
1) Bug fixes:
    a) Pixhawk4 SBUS input and Serial6 pin mapping conflict resolved
------------------------------------------------------------------
Copter 3.6.5-rc2 15-Jan-2019
Changes from 3.6.5-rc1
1) Bug fixes and minor enhancements
    a) RC DSM sync bug fix (channel 8 could temporarily become zero, ChibiOS only)
    b) Pixhawk board LED fix (ChibiOS only)
------------------------------------------------------------------
Copter 3.6.5-rc1 11-Jan-2019
Changes from 3.6.4
1) Bug fixes and minor enhancements
    a) SD Card reliability improvement (BRD_SD_SLOWDOWN param) and allow re-inserting card after boot
    b) Mode and Auxiliary switch range check to protect during FrSky SBUS failsafe recovery
    c) Follow mode reports target distance and heading to ground station
    d) Pixhawk4-mini safety switch and LED fix
    e) Bebop2 build fix
2) RC protocol decoding for SRXL, SUMD and ST24 extended to all boards including pixracer and ChibiOS-only boards
3) DrotekP3 Pro support
4) Cube Purple support
------------------------------------------------------------------
Copter 3.6.4 27-Dec-2018 / 3.6.4-rc1 20-Dec-2018
Changes from 3.6.3
1) Bug fixes and minor enhancements
    a) Detect and clear stuck I2C bus when using ChibiOS
    b) Pixhawk4 fix for battery monitor voltage scaling
    c) TradHeli Dual (aka Chinook) fix for scaling of second swashplate movement
    d) Omnibus F7 IMU orientation fix
    e) Gimbal fix to avoid extreme movement if pilot's transmitter is off at startup 
    f) Follow mode fix to obey FOLL_SYSID parameter (controls which vehicle to follow)
2) Cheerson CX-OF optical flow sensor support
------------------------------------------------------------------
Copter 3.6.3 04-Dec-2018 / 3.6.3-rc1 01-Dec-2018
Changes from 3.6.2
1) Bug fixes and minor enhancements
    a) Pixhawk4 RC input fix
    b) TradHeli parameter default fixes for Loiter and PosHold
    c) Onboard OSD fix for MatekSys 405
    d) Solo power off button delay time increased
------------------------------------------------------------------
Copter 3.6.2 24-Nov-2018 / 3.6.2-rc4 22-Nov-2018
Changes from 3.6.2-rc3
1) Bug fixes and minor changes:
    a) Benewake TFmini short range scaling fix
    b) Fix alt drop after quick switch from RTL to stabilize to RTL
    c) Winch removed from build by default
------------------------------------------------------------------
Copter 3.6.2-rc3 19-Nov-2018
Changes from 3.6.2-rc2
1) Bug fixes:
    a) Benewake TFmini and TF02 driver reliability fix
    b) Spektrum RC parsing fix when more than 7 channels
2) remove bootloader from px4-v2 builds (reduces firmware size)
------------------------------------------------------------------
Copter 3.6.2-rc1/rc2 15-Nov-2018
Changes from 3.6.1
1) Bug fixes:
    a) Benewake TFmini and TF02 driver checksum fix (was missing many sensor readings)
    b) Range finders report healthy to GCS when out-of-range
    c) RPM sensor reliability fix by initialising analog input pin
------------------------------------------------------------------
Copter 3.6.1 10-Nov-2018 / 3.6.1-rc1 06-Nov-2018
Changes from 3.6.0
1) Garmin LidarLite V3HP support 
2) VFR HUD messages send relative altitude if DEV_OPTIONS = 2.  Useful for older OSDs and GCSs
3) Bug fixes:
    a) Battery failsafe voltage parameter conversion fix
    b) Safety switch startup fix (was occasionally not initialised properly)
    c) Benewake TFmini and TF02 driver discards distances over 327m (avoids reporting 655m when can't read distance)
    d) Dataflash erase only availabled when disarmed (avoids crash if user attempted to erase logs while flying)
4) ChibiOS fixes and enhancements:
    a) Pixracer LEDs colours fixed
    b) Terrain support fixed on Pixracer, MindPx-v2, Radiolink mini-pix
    c) RC input processing fix to avoid memory corruption in some rare cases
    d) FuriousFPV F-35 Lightning board support
    e) SpeedyBee F4 board support
    f) Bootloaders for OmnibusF4v6, mRoX2.1-777, Radiolink mini-pix 
    g) Revo-mini support for external barometer
    h) Pins numbers made consistent across boards (setup of some features now more consistent across boards)
    i) enable safety switch on Pixhawk family f7 boards
------------------------------------------------------------------
Copter 3.6.0 26-Oct-2018
Changes from 3.6.0-rc12
1) Bug Fixes:
    a) Object avoidance fix when using 360 Lidar (mini fence was not including first distance)
    b) Flowhold mode descending fix
2) ChibiOS related fixes:
    a) Kakutef7 support pin mapping fixes
    b) PixRacer LED fix
    c) I2C Lidar driver fix to avoid freeze at startup if sensor not connected
------------------------------------------------------------------
Copter 3.6.0-rc12 05-Oct-2018
Changes from 3.6.0-rc11
1) Bug Fixes:
    a) DSHot ESC driver race condition fix (could result in board freezing up)
    b) EKF fix to rounding off home altitude (could result in slightly inaccurate home altitude)
    c) Pixracer safety switch fix (BRD_SAFETYOPTION parameter was handled incorrectly)
------------------------------------------------------------------
Copter 3.6.0-rc11 01-Oct-2018
Changes from 3.6.0-rc10
1) Bug Fixes:
    a) PosHold fix to initialise desired accelerations
    b) Automatic compass orientation fix to account for equivalent rotations
    c) ChibiOS fix to RSSI voltage reading
------------------------------------------------------------------
Copter 3.6.0-rc10 15-Sep-2018
Changes from 3.6.0-rc9
1) Arming/disarming from transmitter can be disabled with ARMING_RUDDER parameter
2) ChibiOS fixes:
    a) UART baud rate rounding fix (allow 921600 baud to work)
    b) ADC fix to allow more simultaneous channels
3) Bug Fixes:

src/Driver/gpio.h

#ifndef __NAVIO_GPIO_H__
#define __NAVIO_GPIO_H__

#include <stdint.h>

namespace Navio {

/* Raspberry Pi GPIO mapping */
#define RPI_GPIO_2    2    // Pin 3     SDA
#define RPI_GPIO_3    3    // Pin 5     SCL
#define RPI_GPIO_4    4    // Pin 7             NAVIO_PPM_INPUT
#define RPI_GPIO_6    6
#define RPI_GPIO_7    7    // Pin 26    CE1     NAVIO_MPU9250_CS
#define RPI_GPIO_8    8    // Pin 24    CE0     NAVIO_UBLOX_CS
#define RPI_GPIO_9    9    // Pin 21    MISO
#define RPI_GPIO_10   10   // Pin 19    MOSI
#define RPI_GPIO_11   11   // Pin 23    SCLK
#define RPI_GPIO_14   14   // Pin 8     TxD
#define RPI_GPIO_15   15   // Pin 10    RxD
#define RPI_GPIO_17   17   // Pin 11            NAVIO_UART_PORT_5
#define RPI_GPIO_18   18   // Pin 12            NAVIO_UART_PORT_4
#define RPI_GPIO_22   22   // Pin 15            NAVIO_UBLOX_PPS
#define RPI_GPIO_23   23   // Pin 16            NAVIO_MPU9250_DRDY
#define RPI_GPIO_24   24   // Pin 18            NAVIO_SPI_PORT_6
#define RPI_GPIO_25   25   // Pin 22            NAVIO_SPI_PORT_5
#define RPI_GPIO_27   27   // Pin 13            NAVIO_PCA9685_OE
#define RPI_GPIO_28   28   // Pin 3
#define RPI_GPIO_29   29   // Pin 4
#define RPI_GPIO_30   30   // Pin 5
#define RPI_GPIO_31   31   // Pin 6

    class Pin {
    public:
        typedef enum {
            GpioModeInput,
            GpioModeOutput
        } GpioMode;

        Pin(uint8_t pin);
        ~Pin();

        bool    init();
        void    setMode(GpioMode mode);
        uint8_t read() const;
        void    write(uint8_t value);
        void    toggle();

    private:
        int getRaspberryPiVersion() const;
        Pin (const Pin&);
        Pin& operator=(const Pin&);

        uint8_t _pin;
        volatile uint32_t *_gpio;
        GpioMode _mode;

        bool    _deinit();
    };

}

#endif // __NAVIO_GPIO_H__
